libcaf_core/test/async/spsc_buffer.cpp

```diff
@@ -180,4 +214,322 @@ SCENARIO("SPSC buffers moves data between actors") {
+        run();
+        CHECK(on_complete_called);
+        CHECK(outputs.empty());
+      }
+    }
+  }
+}
+
+SCENARIO("SPSC buffers drop data when discarding the read end") {
+  GIVEN("an SPSC buffer resource") {
+    WHEN("destroying the read end before adding a publisher") {
+      THEN("the flow of the writing actor gets canceled") {
+        using actor_t = event_based_actor;
+        auto outputs = std::vector<int>{};
+        {
+          auto [rd, wr] = async::make_spsc_buffer_resource<int>(6, 2);
+          sys.spawn([wr{wr}](actor_t* src) {
+            src
+              ->make_observable() //
+              .iota(1)
+              .subscribe(wr);
+          });
+        }
+        // At scope exit, `rd` gets destroyed, closing the buffer.
+        run();
+        CHECK(outputs.empty());
+      }
+    }
+    WHEN("destroying the read end before adding a publisher") {
+      THEN("the flow of the writing actor gets canceled") {
+        using actor_t = event_based_actor;
+        auto outputs = std::vector<int>{};
+        {
+          auto [rd, wr] = async::make_spsc_buffer_resource<int>(6, 2);
+          sys.spawn([wr{wr}](actor_t* src) {
+            src
+              ->make_observable() //
+              .iota(1)
+              .subscribe(wr);
+          });
+          // Only difference to before: have the actor adding an observer that
+          // writes to `wr` before destroying `rd`.
+          run();
+        }
+        // At scope exit, `rd` gets destroyed, closing the buffer.
+        run();
+        CHECK(outputs.empty());
+      }
+    }
+    WHEN("canceling the read end before adding a publisher") {
+      THEN("the flow of the writing actor gets canceled") {
+        using actor_t = event_based_actor;
+        auto outputs = std::vector<int>{};
+        auto [rd, wr] = async::make_spsc_buffer_resource<int>(6, 2);
+        sys.spawn([wr{wr}](actor_t* src) {
+          src
+            ->make_observable() //
+            .iota(1)
+            .subscribe(wr);
+        });
+        rd.cancel();
+        rd.cancel(); // Calling twice must have no side effect.
+        run();
+        CHECK(outputs.empty());
+      }
+    }
+  }
+}
+
+SCENARIO("resources are invalid after calling try_open") {
+  GIVEN("a producer resource") {
+    WHEN("opening it twice") {
+      THEN("the second try_open fails") {
+        auto [rd, wr] = async::make_spsc_buffer_resource<int>(6, 2);
+        CHECK(rd);
+        CHECK_NE(rd.try_open(), nullptr);
+        CHECK(!rd);
+        CHECK_EQ(rd.try_open(), nullptr);
+        auto rd2 = async::consumer_resource<int>{};
+        rd2 = rd;
+        CHECK(!rd2);
+        rd2 = std::move(rd);
+        CHECK(!rd2);
+        CHECK(wr);
+        CHECK_NE(wr.try_open(), nullptr);
+        CHECK(!wr);
+        CHECK_EQ(wr.try_open(), nullptr);
+        auto wr2 = async::producer_resource<int>{};
+        wr2 = wr;
+        CHECK(!wr2);
+        wr2 = std::move(wr);
+        CHECK(!wr2);
```

Comment:
This not only calls `try_open` twice but checks the copy and move assignment as well?